server/analyzer/add_implicit_prefix_lengths.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/analyzer"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
+
+	"github.com/dolthub/doltgresql/server/types"
+)
+
+// defaultIndexPrefixLength is the index prefix length that this analyzer rule applies automatically to TEXT columns
+// in secondary indexes.
+const defaultIndexPrefixLength = 255
```

Comment:
Should we perhaps have a much, much larger prefix? Does Postgres have some kind of implicit prefix-thing for their string types?

From a quick search it seems like a `UNIQUE` index checks across the entire string, so having a prefix means that we'll eventually run into a situation with divergent behavior (and changing this later for pre-existing repos may take a bit of effort). 1023 seems a bit better, 65535 seems like it may cover the majority of strings that we'd run into. I've no clue on the performance or implementation limitations of larger prefixes though.

Reply:
I'll experiment with some different sized values for the default index prefix length. Our main constraint on the prefix size is Dolt's 65kb row size limit, but I agree a larger value seems like it's going to perform better. 

Postgres doesn't have a concept of prefix lengths – we're just applying that here to get some initial support for `TEXT` columns in secondary indexes. Postgres will store the `TEXT` values in-line, unless they are larger than a certain size, and then they get pushed out to a different location on disk and the row contains a ref to that. [Here are the Postgres docs for this technique, which they call TOAST](https://www.postgresql.org/docs/current/storage-toast.html). 

Longer term, I think we'll want to consider something similar. We'll need secondary indexes and primary keys to be able to follow a ref to load a value from an address, otherwise we won't be able to support the same indexes that Postgres can. I think one of the biggest problems in allowing secondary indexes to contain address refs directly is that `TupleComparator` doesn't have access to take a NodeStore to lookup an address, so it can't sort rows correctly. I took a quick pass at this, but because of package import cycles, it looks like it's going to take a bigger refactoring to resolve that. This would obviously have a performance impact to look up addresses for each row, too. Having a new encoding where we could keep the data in-line when it's small and ref it out when it's large (similar to what Postgres does) could be interesting to try out.

Reply:
Dug into this some more... turns out the bigger constraint is actually a MySQL limit on prefix length (768 bytes) that we enforce in Dolt, so I've changed the default implicit prefix length to match that. If we **do** need to change this in the future, customers can update by dropping the index and re-adding it. 

Another optimization for the future is to not use a prefix length for _unique_ indexes, and rely on comparing the hash. We have support for this in Dolt already, but it looks like it's not working properly in Doltgres when that index is used for lookups (since the index contains the address but the real content is being searched for). If that's fixed, it'd be very easy to add a couple lines and optimize this solution a little bit more.

Reply:
That `UNIQUE`-specific change seems like quite the smart way to do things! It's something to keep in mind. That all makes sense though!